chemicalx/models/gcnbmp.py

```diff
@@ -1,14 +1,255 @@
+
+    def forward(self, input_rep, final_rep, graph_index):
+        """
+        Compute an attention-based readout using the input and output layers of the 
+        RGCN encoder for one molecule.
+
+        Parameters:
+            input_rep (Tensor): input nodes representations
+            final_rep (Tensor): final nodes representations
+            graph_index (Tensor): node to graph readout index
+
+        Returns:
+            g (Tensor): graph-level representation
+        """
+        att = torch.sigmoid(self.lin(torch.cat((input_rep, final_rep), 1)))
+        g = att.mul(self.last_rep(final_rep))
+        g = scatter_add(g, graph_index, dim=0)
+        return g
+
+
+class GCNBMPEncoder(nn.Module, core.Configurable):
+    def __init__(
+        self,
+        input_dim,
+        hidden_dims,
+        num_relation,
+        edge_input_dim=None,
+        short_cut=False,
+        batch_norm=False,
+        activation="sigmoid",
+        concat_hidden=False,
+    ):
+        """Instantiate the GCN-BMP encoder.
+        :param input_dim (int): input dimension.
+        :param hidden_dims (list of int): hidden dimensions.
+        :param output_dim (int): output dimension.
+        :param num_relation (int): number of relations.
+        :param edge_input_dim (int, optional): dimension of edge features.
+        :param batch_norm (bool, optional): apply batch normalization on nodes or not.
+        :param activation (str or function, optional): activation function.
+        """
+        super(GCNBMPEncoder, self).__init__()
+
+        if not isinstance(hidden_dims, Sequence):
+            hidden_dims = [hidden_dims]
+        self.input_dim = input_dim
+        self.output_dim = hidden_dims[-1] * (len(hidden_dims) if concat_hidden else 1)
+        self.dims = [input_dim] + list(hidden_dims)
+        self.num_relation = num_relation
+        self.short_cut = short_cut
+        self.concat_hidden = concat_hidden
+
+        self.layers = nn.ModuleList()
+        for i in range(len(self.dims) - 1):
+            self.layers.append(
+                layers.RelationalGraphConv(
+                    self.dims[i], self.dims[i + 1], num_relation, edge_input_dim, batch_norm, activation
+                )
+            )
+            self.layers.append(Highway(self.dims[i + 1], self.dims[i]))
+
+    def forward(self, graph, input):
+        """
+        Compute the node representations and the graph representation(s).
+
+        Require the graph(s) to have the same number of relations as this module.
+
+        Parameters:
+            graph (Graph): :math:`n` graph(s)
+            input (Tensor): input node representations
+
+        Returns:
+            dict with ``node_feature`` field:
+                node representations of shape :math:`(|V|, d)`
+        """
+        hiddens = []
+        layer_input = input
+        prev_gcn = input
+
+        for i, layer in enumerate(self.layers):
```

Comment:
use `from more_itertools import chunked` to do both operations together
```suggestion
       for conv, highway in chunked(self.layers, 2):
```

Reply:
Thanks, didn't know about this one. Will also add it to the dependencies